lib/internal/modules/ts/typescript.js

```diff
@@ -0,0 +1,70 @@
+'use strict';
+
+const {
+  ArrayPrototypeSome,
+  String,
+  StringPrototypeEndsWith,
+} = primordials;
+const { transformSync } = require('internal/deps/swc/wasm');
+const {
+  stringify,
+} = require('internal/modules/helpers');
+const internalFsBinding = internalBinding('fs');
+const {
+  ERR_UNSUPPORTED_DIR_IMPORT,
+} = require('internal/errors').codes;
+const { fileURLToPath, URL } = require('internal/url');
+const { toNamespacedPath } = require('path');
+
+async function load(url, context, nextLoad) {
+  const isTS = context.format === 'typescript';
+  if (isTS) { context = { __proto__: context, format: 'module' }; }
+  const next = await nextLoad(url, context);
+  if (isTS) {
+    const { code } = transformSync(stringify(next.source), {
+      swcrc: false,
+      jsc: {
+        parser: {
+          syntax: 'typescript',
+        },
+        target: 'esnext',
+      },
+    });
+    return {
+      __proto__: null,
+      format: 'module',
+      source: code,
+    };
+  }
+  return next;
+}
+
+async function resolve(specifier, context, nextResolve) {
+  // Check if the file has a known extension
+  const extensions = ['.ts', '.cjs', '.mjs', '.js'];
+  const hasKnownExtension = ArrayPrototypeSome(extensions, (ext) => StringPrototypeEndsWith(specifier, ext));
+  if (!hasKnownExtension) {
+    // If the file doesn't have an extension
+    // it might be missing, lets try to check if its a typescript file
+    // if not move to the next resolver
+    const { parentURL = null } = context;
+    const path = fileURLToPath(parentURL ?
+      new URL(specifier, parentURL).href :
+      new URL(specifier).href);
+    const tsPath = `${path}.ts`;
+    // TODO(marco-ippolito): move this check to c++
+    const stats = internalFsBinding.internalModuleStat(toNamespacedPath(tsPath));
+    switch (stats) {
+      case 0:
+        // File exists, handle it as '.ts'
+        return nextResolve(tsPath, context);
+      case 1:
+        throw new ERR_UNSUPPORTED_DIR_IMPORT(path, fileURLToPath(specifier), String(tsPath));
+      default:
+        break;
+    }
+  }
+  return nextResolve(specifier, context);
+}
+
+module.exports = { load, resolve };
```

Comment:
We should use a null-prototype object to mimic an ESM `Module` record.
```suggestion
module.exports = { __proto__: null, load, resolve };
```